### ISSUE TITLE
photo: CUDA_BruteForceNonLocalMeans.Regression fails with small eps over

### ISSUE THREAD

**Issue by tomoaki0705** [opencv/opencv#17024]
##### System information (version)

- OpenCV => recent master (499e20d) 
- Operating System / Platform => Jetson TX1/Jetson Nano
- Compiler => GCC 5.4.0/GCC 7.4.0

##### Detailed description

Test `CUDA_BruteForceNonLocalMeans.Regression` fails slightly above the eps.

```
penCV version: 4.3.0-dev
OpenCV VCS version: 4.3.0-50-g499e20d
Build type: Release
Compiler: /usr/bin/c++  (ver 5.4.0)

[----------] 1 test from CUDA_BruteForceNonLocalMeans
[ RUN      ] CUDA_BruteForceNonLocalMeans.Regression
/media/ubuntu/sd-ubuntu/opencv/modules/photo/test/test_denoising.cuda.cpp:85: Failure
The max difference between matrices "gray_gold" and "dgray" is 1 at (0, 49), which exceeds "1e-4", where "gray_gold" at (0, 49) evaluates to (102), "dgray" at (0, 49) evaluates to (103), "1e-4" evaluates to 0.0001
[  FAILED  ] CUDA_BruteForceNonLocalMeans.Regression (1354 ms)
[----------] 1 test from CUDA_BruteForceNonLocalMeans (1354 ms total)

[----------] Global test environment tear-down
[==========] 69 tests from 34 test cases ran. (105740 ms total)
[  PASSED  ] 68 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] CUDA_BruteForceNonLocalMeans.Regression
```

The original code is [here](https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/test/test_denoising.cuda.cpp)

https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/test/test_denoising.cuda.cpp#L58-L71

It's comparing CUDA result against test data
https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/test/test_denoising.cuda.cpp#L78-L85

The implementation has hardly changed for 5+ years

https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/src/cuda/nlm.cu

Since it involves a floating operation, I think it's reasonable to loosen the eps.

https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/test/test_denoising.cuda.cpp#L85

In another word, reverting partial portion of this commit https://github.com/opencv/opencv/commit/2b26094cf567ca02115227e0d115e255be31ae5d#diff-98e14616aface5379b11d0540cacad95

Also, this issue doesn't happen on 3.4 branch

```
CTEST_FULL_OUTPUT
OpenCV version: 3.4.10-dev
OpenCV VCS version: 3.4.10-28-g4d35a46c48
Build type: Release
Compiler: /usr/bin/c++  (ver 7.5.0)
Parallel framework: pthreads (nthreads=4)
CPU features: NEON FP16
OpenCL is disabled
TEST: Skip tests with tags: 'mem_6gb', 'verylong'
 :
[----------] 1 test from CUDA_BruteForceNonLocalMeans
[ RUN      ] CUDA_BruteForceNonLocalMeans.Regression
[       OK ] CUDA_BruteForceNonLocalMeans.Regression (2198 ms)
[----------] 1 test from CUDA_BruteForceNonLocalMeans (2198 ms total)
 :
[----------] Global test environment tear-down
[ SKIPSTAT ] 87 tests skipped
[ SKIPSTAT ] TAG='mem_6gb' skip 1 tests
[ SKIPSTAT ] TAG='verylong' skip 1 tests
[ SKIPSTAT ] TAG='skip_other' skip 85 tests
[==========] 931 tests from 64 test cases ran. (1409 ms total)
[  PASSED  ] 931 tests.
```

I'll send a PR later.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by tomoaki0705** [comment]
I just realized the actual cause.
There is a [color convert](https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/test/test_denoising.cuda.cpp#L67) just before calling `nonLocalMeans`

https://github.com/opencv/opencv/blob/499e20da4fbe1b9e69a3457fc9adb16ef0236b8a/modules/photo/test/test_denoising.cuda.cpp#L67

In master, the color conversion became bit-exact, meaning that the result of color conversion is NOT bit-exact between 3.4 branch and master.
That's the reason why this is failing.
So there are several ways to fix this test failure.
- Porting back bit-exact color conversion to 3.4 branch (which seems too much effort)
- Update corresponding image in opencv_extra
- This PR

I'm open for discussion but loosening threshold seems reasonable to what's happening.

**Comment by alalek** [comment]
Bit-exact color conversions should be landed on both 3.4/master branches (if applicable).
You can try to dump cvtColor results into .png or FileStorage and compare them.

> 1e-4

If result type is 8U (after floating point computations), then eps < 1 doesn't make sense.
`1` is fine in test for 8U image.

**Comment by tomoaki0705** [comment]
>Bit-exact color conversions should be landed on both 3.4/master branches (if applicable).
Right.  I was wrong in that sense.
3.4/master RGB2Gray are bit-exact, but the parameter used for the conversion is different between 3.4/master.
Thus result may differ.

- master
https://github.com/opencv/opencv/blob/8c0b0714e76efef4a8ca2a7c410c60e55c5e9829/modules/imgproc/src/color.simd_helpers.hpp#L15-L26
- 3.4
https://github.com/opencv/opencv/blob/8b541e450b511fde9dd363fa55a30fbb6fc0ace6/modules/imgproc/src/color.simd_helpers.hpp#L15-L22

From my point of view, this PR is ready to merge.
If needed, I can put this PR on 3.4, since eps `1e-4` for `8U` types doesn't make sense.
Please let me know

**Comment by alalek** [comment]
Please rebase test fix to 3.4 branch.

**Comment by asmorkalov** [comment]
@tomoaki0705 I reproduced the issue on my Jetson NANO. Thanks for the patch!
